### PR TITLE
OAuth

### PR DESCRIPTION
**Pull request by lukaknezic**

### CODE REVIEW

**Review by zenled — APPROVED** [review]
It seems okay, just different from how I expected it to be. 

This  Oauth implementation is specific to AWS Cognito (using `amplyfy_auth_cognito`).
I believe the goal was to write a chapter about general Oauth implementation, that works with any Oauth provider (using `flutter_appauth`). 

I would merge this chapter and at some point add another chapter on how to use the general Oauth implementation.

File: oauth/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+### 3. Use for refresh token
+```dart
+Future<void> _refreshTokenRequest() async {
+  // Get refresh token from persistence manager
+  final refreshToken = _tokenPersistenceManager.getRefreshToken();
+    
+  final authRepository = null; // Get auth repository
+  final response = await authRepository.refreshToken(TokenRequest(
```
Comment (zenled) [inline oauth/README.md:38]: Make `AuthRepository.auth` and `AuthRepository.refreshToken` more consistent.
Currently, with `.auth` the `clientId, redirectUrl...` are configured inside the the `AuthRepositoryClass`. 
But with `.refreshToken` the `clientId, redirectUrl...` are configured by the caller of the `AuthRepositoryClass`.

I would expect the `.refreshToken` look something like this:
```dart
Future<TokenResponse?> refreshToken(String refreshToken) {
    final request = TokenRequest(
         '<client_id>',
         '<redirect_url>',
         refreshToken: refreshToken,
         issuer: '<issuer>',
         clientSecret: '<client_secret>',
         scopes: ['<scopes>'],
    );

    return _auth.token(request);
}
```

**Review by zenled — CHANGES_REQUESTED** [review]

File: oauth/oauth.dart [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+class AuthRepository {
+  AuthRepository();
+
+  final _auth = const FlutterAppAuth();
+
+  @override
+  Future<Authorization?> authorize() async {
+    try {
+      final response = _auth.authorizeAndExchangeCode(
```
Comment (Truba) [inline oauth/oauth.dart:11]: Does this open web view on it's own?

File: oauth/oauth.dart [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        refreshToken: response.refreshToken!,
+        accessToken: response.accessToken!,
+      );
+    } catch (_) {
+      return null; // exception is thrown if user cancels authorization
+    }
+  }
+
+  @override
+  Future<Authorization?> refreshToken(String refreshToken) async {
```
Comment (Truba) [inline oauth/oauth.dart:31]: Do we use this in some intercepter?
If so can we add that interceptor to here as well?

Comment (zenled) [reply oauth/oauth.dart:31]: This is used in the `AuthorizationInterceptor` to refresh the expired token.
I would avoid adding the `AutorizationInterceptor` to this chapter, as this chapter should only teach how to use OAuth.
The `AutorizationInterceptor` is quite complex. Refreshing the token is only a small part of what it does.

File: oauth/oauth.dart (hunk 2 above)
Comment (zenled) [reply oauth/oauth.dart:11]: Yes, it automatically opens the web-view.

File: oauth/oauth.dart (hunk 3 above)
Comment (zenled) [reply oauth/oauth.dart]: I added the `AuthorizationInterceptor`.

File: oauth/authorization_interceptor.dart [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  Future<Response> _retryRequest(RequestOptions requestOptions) {
+    return _dio.fetch<dynamic>(requestOptions);
+  }
+}
+
+abstract class TokenStorage {
+  String getRefreshToken();
+
+  String getAccessToken();
+  Future<void> setAccessToken(String accessToken);
```
Comment (Truba) [inline oauth/authorization_interceptor.dart]: Should we also store the new access token?

**Review by Truba — APPROVED** [review]

File: oauth/authorization_interceptor.dart (hunk 4 above)
Comment (zenled) [reply oauth/authorization_interceptor.dart]: I'm not entirely sure what you meant here, but I changed it to make it more clear.
Now we have `AuthStorage` (saves auth to local storage) and `AuthRepository` (communicates with backend).
When refreshing the new tokens will be stored.

Comment (Truba) [reply oauth/authorization_interceptor.dart]: Sry 🤦 , I wanted to ask about storing refresh token but wrote access token :( 
But it seem you got the idea anyway, grate work :)

**Review by Truba — APPROVED** [review]

**Pull request merged**